firedrake/preconditioners/patch.py

```diff
@@ -179,17 +135,19 @@ def matrix_funptr(form, state):
         entity_node_map = op2.Map(iterset,
                                   toset, arity,
                                   values=numpy.zeros(iterset.total_size*arity, dtype=IntType))
-        mat = DenseMat(dofset)
+        mat = LocalMat(dofset)
 
-        arg = mat(op2.INC, (entity_node_map[op2.i[0]],
-                            entity_node_map[op2.i[1]]))
+        arg = mat(op2.INC, (entity_node_map, entity_node_map))
         arg.position = 0
         args.append(arg)
-        statedat = DenseDat(dofset)
-        statearg = statedat(op2.READ, entity_node_map[op2.i[0]])
+        statedat = LocalDat(dofset)
+        state_entity_node_map = op2.Map(iterset,
+                                        toset, arity,
+                                        values=numpy.zeros(iterset.total_size*arity, dtype=IntType))
+        statearg = statedat(op2.READ, state_entity_node_map)
```

Comment:
The new loopy codegen passes maps only once if they are identical (object identity). Because the state arg has a different map to the matrix, we need to make the "same" map again.